pkg/util/volid.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright 2019 Ceph-CSI authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"errors"
+	"strings"
+)
+
+/*
+VolumeIdentifier contains the elements that form a volume ID to be returned by CSI, and contains
+enough information to decompose and extract required cluster and pool information to locate the
+volume that relates to the volume ID.
+
+The volume identifier is composed as elaborated in the comment against ComposeVolID and thus,
+DecomposeVolID is the inverse of the same function.
+
+The VolumeIdentifier structure carries the following fields,
+- PoolID: 32 bit integer of the pool that the volume belongs to, where the ID comes from Ceph pool
+  identifier for the corresponding pool name.
+- EncodingVersion: Carries the version number of the encoding scheme used to encode the volume ID,
+  and is preserved for any future proofing w.r.t changes in the encoding scheme, and to retain
+  ability to parse backward compatible encodings.
+- ClusterID: Is a unique ID per cluster that the CSI instance is serving and is restricted to
+  lengths that can be accommodated in the encoding scheme.
+- ImageName: Is the on-disk image name of the CSI volume that corresponds to this volume ID.
+*/
+type VolumeIdentifier struct {
+	PoolID          uint32
+	EncodingVersion uint16
+	ClusterID       string
+	ImageName       string
+}
+
+// This maximum comes from the CSI spec on max bytes allowed in the VolumeID field
+const maxVolIDLen = 128
+
+/*
+ComposeVolID composes a volume ID from passed in parameters.
+Version 1 of the encoding scheme is as follows,
+	[volume_id_version=1:4byte] + [-:1byte]
+	[length of clusterID=1:4byte] + [-:1byte]
+	[clusterID:36bytes (MAX)] + [-:1byte]
+	[poolID:8bytes] + [-:1byte]
+	[length of ImageName=1:4byte] + [-:1byte]
+	[ImageName: (should not overflow total length of maxVolIDLen)]
```

Comment:
Nit: really should only be the UUID portion and doesn't need to include the ```csi-vol``` / ```csi-snap``` prefix.

Reply:
Ack! as I address a similar comment in naming in the rbd package, will address it here.

Just from a read and use perspective it maybe worthwhile to retain the naming, but that can also be construed as lazy programming. Will attempt to see if it makes sense to remove it without impacting readability or adding too much wrapper code.